### PR TITLE
Fix juju exec

### PR DESCRIPTION
**Pull request by hypeitnow**
improved juju exec usage and installing the recorder

### CODE REVIEW

File: jhack/utils/simulate_event.py [hunk 1]
```diff
@@ -126,8 +126,8 @@ def _simulate_event(
         override=env_override,
         operator_dispatch=operator_dispatch,
     )
-    # todo: insert `sudo` if this is a machine unit!
-    cmd = f"juju ssh {unit} /usr/bin/{_J_EXEC_CMD} -u {unit} {env} ./dispatch"
+    # todo: insert `sudo` if this is a machine unit! - this is now done automatically
```
Comment (PietroPasotti) [inline jhack/utils/simulate_event.py:129]: see https://github.com/PietroPasotti/jhack/pull/40 which is already fixing the same issue. Feel free to give a review there!

IMHO it's cleaner to only use sudo when we know it's needed. On containers we don't.

File: jhack/utils/simulate_event.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 from jhack.logger import logger as jhack_logger
 
 # note juju-exec is juju-run in juju<3.0
-_J_EXEC_CMD = "juju-exec" if juju_agent_version() >= (3, 0) else "juju-run"
+_J_EXEC_CMD = "juju exec" if juju_agent_version() >= (3, 0) else "juju run"
```
Comment (PietroPasotti) [inline jhack/utils/simulate_event.py:9]: why this change? did you see errors?
Note that this command is not run on the juju cli, but in the unit. Is there a 'juju' binary in the unit?

File: jhack/utils/event_recorder/client.py [hunk 3]
```diff
@@ -322,6 +335,8 @@ def _install(unit: str):
         print("Alright.")
 
     print("Shelling over recorder script...")
+    _ensure_recorder_dir(unit)
```
Comment (PietroPasotti) [inline jhack/utils/event_recorder/client.py:338]: in what situations could this path not exist?

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
Many thanks for the contribution!
A few things are unclear. Also this PR seems to address multiple issues at once (related to jhack on machine charms). Would appreciate a few words on each change and what it's trying to solve.

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -70,4 +70,5 @@ plugs:
   # read access to $HOME. Required for sync and other commands that manipulate
   #  your local code projects.
   home-read:
+    allow-other-read: true
```
[the remark this answers is not in the document]
Comment (hypeitnow) [reply snap/snapcraft.yaml]: I was using outdated documentation, the attempt was to allow snap to read other users' direcotries but it seems to be resolved in https://github.com/PietroPasotti/jhack/issues/38#issuecomment-1385065278

File: jhack/utils/simulate_event.py (hunk 2 above)
Comment (hypeitnow) [reply jhack/utils/simulate_event.py:9]: I reanalysed the code and indeed it works this way. Do you have any article in which you explain why it's better to run juju-exec from within the unit?

File: jhack/utils/event_recorder/client.py (hunk 3 above)
Comment (hypeitnow) [reply jhack/utils/event_recorder/client.py:338]: I was testing a kubernetes core deployment using vms generated in maas, the directory was not there although I did not change anything after juju agent installation on these machines. The version I used was 3-0-3

File: jhack/utils/simulate_event.py (hunk 1 above)
Comment (hypeitnow) [reply jhack/utils/simulate_event.py:129]: If the simulate event sticks to running from inside the machine instead of running by juju client then https://github.com/PietroPasotti/jhack/pull/40 is a greate solution to differentiate between root and standard user execution

Comment (hypeitnow) [reply jhack/utils/simulate_event.py:129]: The latest juju documentation mentions automatic usage of sudo in case of running juju exec from client for a unit and using ubuntu user in case of application usage.

Comment (PietroPasotti) [reply jhack/utils/simulate_event.py]: simulate event does 'stick to running from inside the machine' as you mention. it does not run `juju exec`, instead, it runs `juju ssh <unit> /path/to/juju-exec`

File: jhack/utils/event_recorder/client.py (hunk 3 above)
Comment (PietroPasotti) [reply jhack/utils/event_recorder/client.py:338]: I mean, the recorder dir is the directory where the charm.py file is (`src`). I don't know why you don't have a src folder in the first place, but my point is: without a src folder (and thus without a charm.py file), what are we even doing with the recorder?

I must be missing something essential here... maybe we should have a call to discuss?
